### PR TITLE
[BBC] BBC Pluggable ImageProcessors and the introduction of ImageProcessorResources

### PR DESCRIPTION
**Pull request by gribeiro**
## What does this change?

1. Makes an actor system available to processors so that updates can be scheduled for stores.
2. Rationalises the way the processor local and application/module config is passed through.
3. This is done by introducing a `ImageProcessorResources` type that includes the processor config, the common config and the actor system for injection into dynamically loaded processors.
4. Making the play actor system available in the config classes so it can be injected for the above.
5. Update with a first version of the BBC's metadata image processors.

## How can success be measured?

We need to ensure that the changes introduced in item 1 do not impact the current application Image Processor and all the Image Processors still works with this change.

## Tested?
- [ x] locally
- [ ] on TEST

### CODE REVIEW

**Comment by sihil** [comment]
Thanks for this @gribeiro - I'll take a look shortly.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/components/BBCImageProcessorsDependencies.scala
Comment (sihil) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/components/BBCImageProcessorsDependencies.scala:12]: For me this memoisation logic could be confusing (were multiple configurations to be used for different processors, this will result in only the first one being initialised actually working). Having `save` be a Map[A, B] and only computing the output once for each different input might resolve this, assuming A correctly implements `Comparable`. 

That said this is logic that won't be used outside the BBC so I don't consider it as a blocker for merging.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/components/BBCMetadataStore.scala
Comment (sihil) [inline common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/components/BBCMetadataStore.scala:12]: Might be best for consistency to name this class `BBCMetadataStore` to match the file.

File: image-loader/app/lib/ImageLoaderConfig.scala
Comment (sihil) [inline image-loader/app/lib/ImageLoaderConfig.scala:57]: I'd advocate passing in CommonConfig rather than the plain play config (so passing in `this` here).

File: image-loader/test/scala/lib/ImageProcessorLoaderTest.scala
Comment (sihil) [inline image-loader/test/scala/lib/ImageProcessorLoaderTest.scala:77]: I think both of these tests are useful to have and can be retained more easily be introducing a resources type.

File: image-loader/test/scala/lib/ImageProcessorLoaderTest.scala
Comment (sihil) [inline image-loader/test/scala/lib/ImageProcessorLoaderTest.scala:160]: Likewise - this is a useful test as it helps ensure that an image processor cannot easily be mis-configured.

**Review by sihil — CHANGES_REQUESTED** [review]
Hi @gribeiro - thanks for this PR.

It's great to see the pluggable image processor stuff being used and it's clear that you've hit the limitations that I imagined. I think you've made good steps to remove those limitations but it would be good to expand it and make it generally more powerful.

Rather than overload the single play `Configuration` argument with both the global/module config and also the processor config - I'd suggest introducing an `ImageProcessorResources` type that can carry multiple configurations and resources separately and use that as the single argument to image processors. At this stage we don't need to be backwards compatible so can just change it.

That also means we can include an actor system, which is required to schedule updates to the metadata store (if indeed you want them to dynamically update rather than update at startup only).

Getting an actor system is a bit non-trivial and will need a change to the way the configuration is handled in `GridComponents`.

I don't see the requirement to remove the tests as they are covering cases that still make sense to me. Likewise the assertion they rely on still makes sense to me. Please do make an argument for their removal if you disagree.

I've made a quick PR against your branch to sketch out some improvements. Happy to pair on hangout/zoom if you want to run through anything.

https://github.com/bbc/grid/pull/61

File: common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/components/BBCImageProcessorsDependencies.scala
Comment (gribeiro) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/components/BBCImageProcessorsDependencies.scala:14]: I agree, it makes sense if only it is used with the CommonConfiguration. The first version was a Map, but it has the potential to have some memory leak. For instance, if you let hot reload the config many times.
I will change the signature from ` ImageProcessorResources => MetadataStore` to `CommonConfig =>MetadataStore ` in order to be less confusing.

**Comment by gribeiro** [comment]
Hi @sihil. 

I merged all the suggestions and did some minor changes and fixes.

1. Changed the name of the class MetadataStore to BBCMetadataStore.
2. Small fix on loadImageProcessor -- the implementation of actorSystem.
3. Created a BBCDependenciesConfig with the configuration that should be loaded only once.

Thank you :)

**Review by sihil — APPROVED** [review]
This is great work. I've left a couple of minor comments that you might want to address so I won't merge this until you give the go ahead - but I will test it in our integration environment so we are ready to merge when you're happy.

**Comment by sihil** [comment]
@gribeiro We've tested this in our test environment and are happy to merge this when you're ready.

File: common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/BBCSupplierProcessors.scala
Comment (gribeiro) [reply common-lib/src/main/scala/com/gu/mediaservice/lib/bbc/BBCSupplierProcessors.scala:34]: Yes. We will introduce more changes, we need this first version to do some testing first.

**Comment by gribeiro** [comment]
Hello @sihil. Thank you for all the great suggestions.

I update a comment and added the descriptions to the ImageProcessors.
I also update the description and the title of the PR to better reflect our work.

Now the checks are failing on `lib.elasticsearch.SyndicationRightsOpsTest`.
I'm not sure what is happening. Do you have any idea what this might be?

**Pull request merged**
